src/Clay/Grid.hs

```diff
@@ -56,3 +109,669 @@ gridGap = key "grid-gap"
+  --
+  -- __NOTE:__ negative 'Integer' or 0 are invalid.
+  | Span (Maybe CustomIdentGrid) (Maybe Integer)
+
+  -- | @auto@ CSS keyword.
+  | Auto
+
+  -- | @inherit@ CSS keyword.
+  | Inherit
+
+  -- | @initial@ CSS keyword.
+  | Initial
+
+  -- | @unset@ CSS keyword.
+  | Unset
+  deriving (Eq, Show)
+
+class ToGridLine a where
+  -- | Convert the provided type to a 'GridLine'.
+  toGridLine :: a -> GridLine
+
+instance ToGridLine GridLine where
+  toGridLine = id
+
+instance ToGridLine Integer where
+  -- | __NOTE:__ 'Integer' value of 0 is invalid.
+  toGridLine = Coordinate
+
+-- | @custom-ident@ value.
+instance ToGridLine CustomIdentGrid where
+  toGridLine x = GridLineCustomIdent x Nothing
+
+-- | @custom-ident@ value.
+instance ToGridLine String where
+  toGridLine = toGridLine . partialMkCustomIdentGrid . T.pack
+
+-- | Both @custom-ident@ and `Integer` values, provided as a pair.
+--
+-- __NOTE:__ 'Integer' value of 0 is invalid.
+instance ToGridLine (CustomIdentGrid, Integer) where
+  toGridLine (x, y) = GridLineCustomIdent x (Just y)
+
+-- | Both @custom-ident@ and `Integer` values, provided as a pair.
+--
+-- __NOTE:__ 'Integer' value of 0 is invalid.
+instance ToGridLine (String, Integer) where
+  toGridLine (x, y) = toGridLine (partialMkCustomIdentGrid $ T.pack x, y)
+
+-- | One or two @grid-line@ values.
+data GridLines2
+    -- | One @grid-line@ value.
+  = One2 OneGridLine
+    -- | Two @grid-line@ values.
+  | Two2 TwoGridLines
+
+class ToGridLines2 a where
+
+  -- | Convert the provided type to 'GridLines2' (one or two @grid-line@ values).
+  toGridLines2 :: a -> GridLines2
+
+-- | One @grid-line@ value.
+instance ToGridLines2 GridLine where
+  toGridLines2 = One2 . OneGridLine
+
+-- | One @grid-line@ value.
+instance ToGridLines2 OneGridLine where
+  toGridLines2 = One2
+
+-- | Two @grid-line@ values.
+instance ToGridLines2 TwoGridLines where
+  toGridLines2 = Two2
+
+-- | One or two @grid-line@ values.
+instance ToGridLines2 GridLines2 where
+  toGridLines2 = id
+
+-- | One 'Integer' value.
+--
+-- __NOTE:__ 'Integer' value of 0 is invalid.
+instance ToGridLines2 Integer where
+  toGridLines2 = toGridLines2 . toGridLine
+
+-- | One @custom-ident@ value.
+instance ToGridLines2 CustomIdentGrid where
+  toGridLines2 = toGridLines2 . toGridLine
+
+-- | One @custom-ident@ value.
+instance ToGridLines2 String where
+  toGridLines2 = toGridLines2 . toGridLine
+
+-- | One time both a @custom-ident@ and 'Integer' values, provided as a pair.
+--
+-- __NOTE:__ 'Integer' value of 0 is invalid.
+instance ToGridLines2 (CustomIdentGrid, Integer) where
+  toGridLines2 = toGridLines2 . toGridLine
+
+-- | One time both a @custom-ident@ and 'Integer' values, provided as a pair.
+--
+-- __NOTE:__ 'Integer' value of 0 is invalid.
+instance ToGridLines2 (String, Integer) where
+  toGridLines2 = toGridLines2 . toGridLine
+
+-- | One, two, three or four @grid-line@ values.
```

Comment:
I'm surprised, is there a maximum number of 4 grid lines in CSS?

Reply:
It doesn't refer to the number of grid lines. `<grid line>` values are a size or position argument on the grid. See [grid-area](https://developer.mozilla.org/en-US/docs/Web/CSS/grid-area).

`grid-row` and `grid-column` properties may have 1 or 2 `<grid line>` values. `grid-area` may have 1, 2, 3 or 4.

The `ToGridLines4` type class is used by one of the instances of the `Slash` type class (lines 611-612).

The `//` operator let you use 1, 2, 3 or 4 arguments of type `GridLine` for the `gridArea` function (and only 1 or 2 for the `gridRow` or `gridColumn` ones). See the various tests in the Spec file (lines 389-451 for `gridArea`).